controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -673,11 +977,63 @@ func diffAddresses(oldAddresses, newAddresses AddressSet) (add, remove AddressSe
 		}
 	}
 	add = AddressSet{
-		Addresses: addAddesses,
+		Addresses: addAddresses,
 		Labels:    newAddresses.Labels,
 	}
 	remove = AddressSet{
 		Addresses: removeAddresses,
 	}
 	return
 }
+
+func isTargetPortInSlice(targetPort namedPort, slice *discovery.EndpointSlice) bool {
+	if slice.Ports == nil && slice.Endpoints == nil {
```

Comment:
Sort of an edge case, let:
`web` -- a service
`pod-1` -- web-pod-1
`pod-2` -- web-pod-2
`s` -- slice for `web`: {`pod-1-IP`, `pod-2-IP`}

When we execute
`kubectl delete pod-1, pod-2`

Then we expect `NoEndpoints` for that slice. In actuality, if we don't include this check, we will be left with an endpoint that does not exist, but the application thinks exists.
1. Pod-1 gets deleted first -- s updates itself (removes pod-1-ip) --> dest-service sends a remove for pod-1-ip
2.1. Pod-2 gets deleted second -- s updates itself (removes pod-2-ip) --> s now has NIL PORT and NIL Endpoints => `isTargetPortInSlice` returns false (without this line) since the port is nil, the target port of the service is definitely not in there.
2.2. Pod-2 gets deleted second -- s updates itself (removes pod-2-ip) --> s now has NIL PORT and NIL Endpoints => `isTargetPortInSlice` returns true, the slice is processed, dest-service sends a remove for pod-2-ip.

Hope it makes sense.

Reply:
I think if we push the port handling logic down into `resolveESTargetPort` we can avoid these kinds of edge cases.